src/Workspaces/Remote/Core/Serialization/ImmutableCollectionMessagePackResolver.cs

```diff
@@ -0,0 +1,394 @@
+            {
+                return intermediateCollection.ToImmutable();
+            }
+
+            protected override ImmutableDictionary<TKey, TValue>.Builder Create(int count, MessagePackSerializerOptions options)
+            {
+                return ImmutableDictionary.CreateBuilder<TKey, TValue>(options.Security.GetEqualityComparer<TKey>());
+            }
+        }
+
+        internal sealed class InterfaceImmutableSetFormatter<T> : CollectionFormatterBase<T, ImmutableHashSet<T>.Builder, IImmutableSet<T>>
+        {
+            protected override void Add(ImmutableHashSet<T>.Builder collection, int index, T value, MessagePackSerializerOptions options)
+            {
+                collection.Add(value);
+            }
+
+            protected override IImmutableSet<T> Complete(ImmutableHashSet<T>.Builder intermediateCollection)
+            {
+                return intermediateCollection.ToImmutable();
+            }
+
+            protected override ImmutableHashSet<T>.Builder Create(int count, MessagePackSerializerOptions options)
+            {
+                return ImmutableHashSet.CreateBuilder<T>(options.Security.GetEqualityComparer<T>());
+            }
+        }
+
+        internal sealed class InterfaceImmutableQueueFormatter<T> : CollectionFormatterBase<T, ImmutableQueueBuilder<T>, IImmutableQueue<T>>
+        {
+            protected override void Add(ImmutableQueueBuilder<T> collection, int index, T value, MessagePackSerializerOptions options)
+            {
+                collection.Add(value);
+            }
+
+            protected override IImmutableQueue<T> Complete(ImmutableQueueBuilder<T> intermediateCollection)
+            {
+                return intermediateCollection.Q;
+            }
+
+            protected override ImmutableQueueBuilder<T> Create(int count, MessagePackSerializerOptions options)
+            {
+                return new ImmutableQueueBuilder<T>();
+            }
+        }
+
+        internal sealed class InterfaceImmutableStackFormatter<T> : CollectionFormatterBase<T, T[], IImmutableStack<T>>
+        {
+            protected override void Add(T[] collection, int index, T value, MessagePackSerializerOptions options)
+            {
+                collection[collection.Length - 1 - index] = value;
+            }
+
+            protected override IImmutableStack<T> Complete(T[] intermediateCollection)
+            {
+                return ImmutableStack.CreateRange(intermediateCollection);
+            }
+
+            protected override T[] Create(int count, MessagePackSerializerOptions options)
+            {
+                return new T[count];
+            }
+        }
+
+        // pseudo builders
+        internal sealed class ImmutableQueueBuilder<T>
+        {
+            public ImmutableQueue<T> Q { get; set; } = ImmutableQueue<T>.Empty;
+
+            public void Add(T value)
+            {
+                Q = Q.Enqueue(value);
```

Comment:
so what does this mean.  if we deserialize a queue with N elements, we'd just do N adds?

Reply:
See the comment at the beginning of the file.